File: framework/languages/java/samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/application/roomallocation/BingoRoomAllocator.java
package systems.zlink.samples.bingo.server.play.application.roomallocation;

import systems.zlink.samples.bingo.server.play.domain.bingo.BingoRoomModels;

public final class BingoRoomAllocator {
    private final BingoMatchQueue matchQueue;
    private final long drawPeriodMillis;
    private int roomSeq;

    public BingoRoomAllocator(BingoMatchQueue matchQueue, long drawPeriodMillis) {
        this.matchQueue = matchQueue;
        this.drawPeriodMillis = drawPeriodMillis;
    }

    public BingoRoomAllocation allocate(String actorId, String mode) {
        if (actorId == null || actorId.isBlank()) {
            throw new IllegalStateException("actorId is required");
        }
        if (!"two-player".equals(mode)) {
            throw new IllegalStateException("Unsupported bingo mode. mode=" + mode);
        }

        int roomSeq = nextRoomSeq();
        BingoRoomModels.BingoRoomSettings settings =
            BingoRoomModels.BingoRoomSettings.create(mode, roomSeq, drawPeriodMillis);
        BingoMatchReservation reservation = matchQueue.reserve(
            mode,
            actorId,
            settings.mode() + "-room-" + roomSeq,
            settings.requiredPlayers());
        return new BingoRoomAllocation(reservation.roomId(), settings);
    }

    private synchronized int nextRoomSeq() {
        roomSeq += 1;
        return roomSeq;
    }

}

File: framework/languages/java/samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/application/roomallocation/BingoRoomAllocator.kt
package systems.zlink.samples.kotlin.bingo.server.play.application.roomallocation

import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomSettings

class BingoRoomAllocator(
    private val matchQueue: BingoMatchQueue,
    private val drawPeriodMillis: Long,
) {
    private var roomSeq: Int = 0

    suspend fun allocate(
        actorId: String,
        mode: String,
    ): BingoRoomAllocation {
        if (actorId.isBlank()) {
            throw IllegalStateException("actorId is required")
        }
        if (mode != "two-player") {
            throw IllegalStateException("Unsupported bingo mode. mode=$mode")
        }

        val settings = BingoRoomSettings.create(mode, nextRoomSeq(), drawPeriodMillis)
        val reservation = matchQueue.reserve(
            mode,
            actorId,
            "${settings.mode}-room-$roomSeq",
            settings.requiredPlayers,
        )
        return BingoRoomAllocation(reservation.roomId, settings)
    }

    @Synchronized
    private fun nextRoomSeq(): Int {
        roomSeq += 1
        return roomSeq
    }
}
